core/src/main/java/org/infinispan/remoting/transport/jgroups/RspListFuture.java

```diff
@@ -0,0 +1,39 @@
+package org.infinispan.remoting.transport.jgroups;
+
+import org.infinispan.remoting.responses.Response;
+import org.infinispan.util.concurrent.TimeoutException;
+import org.jgroups.blocks.GroupRequest;
+import org.jgroups.util.FutureListener;
+import org.jgroups.util.Rsp;
+import org.jgroups.util.RspList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+/**
+ * @author Dan Berindei
+ * @since 8.0
+ */
+class RspListFuture extends CompletableFuture<RspList<Response>> implements FutureListener<RspList<Response>> {
```

Comment:
Not sure this is a good abstraction for the reasons explained above. 

Each response really comes from a different node, so each is independent IMO, so it's more flexible to have a List of CompletableFutures (if that's doable, as opposed to a CompletableFuture wrapping a list).

Given a list of CompletableFutures, you can wait for all of them or only one of them. With a CompletableFuture of a List, you only have one option which is to wait for the entire list of responses.

I don't know the JGroups API too well but maybe JGroups itself is limiting you to do this?